### PR TITLE
[OPPRO-277] Support chr() and abs() functions in Gluten

### PR DESCRIPTION
**Pull request by zhixingheyi-tian**
## What changes were proposed in this pull request?

Also include String function chr() support.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: jvm/src/main/scala/io/glutenproject/expression/ConverterUtils.scala [hunk 1]
```diff
@@ -341,4 +341,8 @@ object ConverterUtils extends Logging {
   final val STARTS_WITH = "starts_with"
   final val SUBSTRING = "substring"
   final val ROW_CONSTRUCTOR = "row_constructor"
+
+  //
+  final val CHR = "chr"
+  final val ABS = "abs"
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/expression/ConverterUtils.scala:347]: Are these two functions already defined in Substrait yaml? You could check here: https://github.com/substrait-io/substrait/tree/main/extensions. If not, you can open PR to substrait to add.

File: jvm/src/main/scala/io/glutenproject/expression/ConverterUtils.scala [hunk 2]
```diff
@@ -341,4 +341,8 @@ object ConverterUtils extends Logging {
   final val STARTS_WITH = "starts_with"
   final val SUBSTRING = "substring"
   final val ROW_CONSTRUCTOR = "row_constructor"
+
+  //
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/expression/ConverterUtils.scala:345]: What is the content here?

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala [hunk 3]
```diff
@@ -94,7 +94,34 @@ class AbsTransformer(child: Expression, original: Expression)
     with ExpressionTransformer
     with Logging {
   override def doTransform(args: java.lang.Object): ExpressionNode = {
-    throw new UnsupportedOperationException("Not supported: Abs.")
+    val child_node = child.asInstanceOf[ExpressionTransformer].doTransform(args)
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala:97]: Please use Camel-Case name for child_node.

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala [hunk 4]
```diff
@@ -94,7 +94,34 @@ class AbsTransformer(child: Expression, original: Expression)
...
+    if (!child_node.isInstanceOf[ExpressionNode]) {
+      throw new UnsupportedOperationException(s"Not supported yet.")
+    }
+    val functionMap = args.asInstanceOf[java.util.HashMap[String, java.lang.Long]]
+    val functionId = ExpressionBuilder.newScalarFunction(functionMap,
+      ConverterUtils.makeFuncName(ConverterUtils.ABS, Seq(child.dataType), FunctionConfig.OPT))
+    val expressionNodes = Lists.newArrayList(child_node.asInstanceOf[ExpressionNode])
+    val typeNode = TypeBuilder.makeFP64(true)
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala:105]: Should not be hard-coded as fp64 here. Also, please pay attention to the nullability.

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala [hunk 5]
```diff
@@ -94,7 +94,34 @@ class AbsTransformer(child: Expression, original: Expression)
...
+  }
+}
+
+class ChrTransformer(child: Expression, original: Expression)
+  extends Chr(child: Expression)
+    with ExpressionTransformer
+    with Logging {
+  override def doTransform(args: java.lang.Object): ExpressionNode = {
+    val child_node = child.asInstanceOf[ExpressionTransformer].doTransform(args)
+    if (!child_node.isInstanceOf[ExpressionNode]) {
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala:116]: Same.

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala [hunk 6]
```diff
@@ -94,7 +94,34 @@ class AbsTransformer(child: Expression, original: Expression)
...
+      throw new UnsupportedOperationException(s"Not supported yet.")
+    }
+    val functionMap = args.asInstanceOf[java.util.HashMap[String, java.lang.Long]]
+    val functionId = ExpressionBuilder.newScalarFunction(functionMap,
+      ConverterUtils.makeFuncName(ConverterUtils.CHR, Seq(child.dataType), FunctionConfig.OPT))
+    val expressionNodes = Lists.newArrayList(child_node.asInstanceOf[ExpressionNode])
+    val typeNode = TypeBuilder.makeString(true)
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala:123]: The nullabiltity should not be hard-coded.

**Comment by FelixYBW** [comment]
@zhixingheyi-tian can you create an issue in issue list or internal issue tracker?

File: jvm/src/main/scala/io/glutenproject/expression/ConverterUtils.scala (hunk 2 above)
Comment (zhixingheyi-tian) [reply jvm/src/main/scala/io/glutenproject/expression/ConverterUtils.scala]: done

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala (hunk 3 above)
Comment (zhixingheyi-tian) [reply jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala]: Done

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala (hunk 4 above)
Comment (zhixingheyi-tian) [reply jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala]: Done

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala (hunk 5 above)
Comment (zhixingheyi-tian) [reply jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala]: Done

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala (hunk 6 above)
Comment (zhixingheyi-tian) [reply jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala]: Done

**Comment by zhixingheyi-tian** [comment]
> @zhixingheyi-tian can you create an issue in issue list or internal issue tracker?

Have created corresponding Jira. 
Thanks

File: backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala [hunk 7]
```diff
@@ -179,9 +179,38 @@ class TestOperator extends WholeStageTransformerSuite {
...
+    df.explain(false)
+    assert(df.queryExecution.executedPlan.find(_.isInstanceOf[ProjectExecTransformer]).isDefined)
+  }
+
+  test("Test abs function") {
+    val df = spark.sql("SELECT abs(l_orderkey) from lineitem")
+    df.printSchema
+    df.show()
+    df.explain(false)
+    assert(df.queryExecution.executedPlan.find(_.isInstanceOf[ProjectExecTransformer]).isDefined)
```
Comment (rui-mo) [inline backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala:200]: Could you also assert the result?

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala [hunk 8]
```diff
@@ -94,7 +94,44 @@ class AbsTransformer(child: Expression, original: Expression)
...
+    val typeNode = original.dataType match {
+      case DoubleType =>
+        TypeBuilder.makeFP64(nullable)
+      case LongType =>
+        TypeBuilder.makeI64(nullable)
+      case IntegerType =>
+        TypeBuilder.makeI32(nullable)
+      case otherType =>
+        throw new UnsupportedOperationException(s"Type $otherType not supported in abs().")
+    }
```
Comment (rui-mo) [inline jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala:115]: Please use `getTypeNode` function in ConverterUtils to get rid of this.

File: jvm/src/main/scala/io/glutenproject/expression/ConverterUtils.scala (hunk 1 above)
Comment (zhixingheyi-tian) [reply jvm/src/main/scala/io/glutenproject/expression/ConverterUtils.scala:347]: Have created a PR to  substrait-io/substrait

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala (hunk 8 above)
Comment (zhixingheyi-tian) [reply jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala:115]: Because ConverterUtils.getTypeNode is not fit to abs().

Comment (rui-mo) [reply jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala:115]: It is used to convert Spark datatype into Substrait type node. I think above types are already covered by that function.

File: backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala (hunk 7 above)
Comment (zhixingheyi-tian) [reply backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala:206]: Done

File: jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala (hunk 8 above)
Comment (zhixingheyi-tian) [reply jvm/src/main/scala/io/glutenproject/expression/UnaryOperatorTransformer.scala]: Done

**Review by rui-mo — APPROVED** [review]
LGTM.

**Comment by zzcclp** [comment]
Please rebase and re-test.

Also move test case to gluten-ut module ?  After CH Backend supports these expression and reuse test cases.

**Review by zzcclp — APPROVED** [review]
LGTM

**Comment by zhixingheyi-tian** [comment]
> Please rebase and re-test.
> 
> Also move test case to gluten-ut module ? After CH Backend supports these expression and reuse test cases.

Have re-tested.


Similar here: https://github.com/oap-project/gluten/pull/451#discussion_r1002297910. Can in another PR.

**Comment by zhixingheyi-tian** [comment]
> LGTM

Thanks

**Comment by zzcclp** [comment]
> > Please rebase and re-test.
> > Also move test case to gluten-ut module ? After CH Backend supports these expression and reuse test cases.
> 
> Have re-tested.
> 
> Similar here: [#451 (comment)](https://github.com/oap-project/gluten/pull/451#discussion_r1002297910). Can in another PR.

OK, I will do this after supporting these functions for ClickHouse

**Pull request merged**